imports/composite_redolog_manager.go

```diff
@@ -0,0 +1,330 @@
+//  Copyright (c) 2017-2018 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package imports
+
+import (
+	"encoding/json"
+	"errors"
+	"github.com/Shopify/sarama"
+	"github.com/uber/aresdb/memstore/common"
+	metaCom "github.com/uber/aresdb/metastore/common"
+	"github.com/uber/aresdb/utils"
+	"sync"
+)
+
+// CompositeRedologManager is the class to take data ingestion from all data source (kafka, http, etc.), write to local redolog when necessary,
+// and then call storeFunc to store data
+type CompositeRedologManager struct {
+	sync.RWMutex `json:"-"`
+	// table name
+	Table string `json:"table"`
+	// table shard id
+	Shard int `json:"shard"`
+	// pointer back to the factory
+	factory *RedologManagerFactory
+	// Local file redolog manager
+	fileRedoLogManager *FileRedologManager
+	// Kafka consumer if kafka import is supported
+	kafkaReader *kafkaPartitionReader
+	// ready channel is used to indicate if recovery is finished
+	readyChan chan bool
+	// function to store received upsertbatch
+	storeFunc BatchStoreFunc
+
+	// convenient variables
+	recoveryDone       bool
```

Comment:
generally speaking, you don't need a channel and a variable to tell whether  recovery is done. Use one single conditional variable instead